Expand expect module test suite

pkingstonxyz opened the pull request:
##### SUMMARY

Add tests to cover several exception handlers in expect.py
Should bring test coverage for expect.py to 100%

##### ISSUE TYPE

Test Pull Request

pkingstonxyz committed:
Test expect module for missing pexpect

pkingstonxyz committed:
Test expect module for incorrect timeout type

pkingstonxyz committed:
Test expect module for out of date pexpect failures

pkingstonxyz committed:
Test pexpect.ExceptionPexpect

Also fixed task naming to be inline with the rest of the file.

pkingstonxyz committed:
Change pexpect management to venv

pkingstonxyz committed:
Normalize b_out to b'' when falsy for consistent handling

pkingstonxyz committed:
Move test venv from /tmp/ to output_dir

nitzmahone commented on test/integration/targets/expect/tasks/main.yml:
```diff
@@ -241,3 +241,106 @@
+    - name: assert pexpect dependency absence
+      assert:
+        that:
+          - "pexpect_dependency_absence.failed"
+          - "'pexpect' in pexpect_dependency_absence.msg"
```
YAML quoting is "fun" (this coming from a maintainer of PyYAML :laughing:)- most Jinja expression strings (including those that were already in the test) don't actually need outer quotes, or can be trivially rewritten not to:

```suggestion
          - pexpect_dependency_absence.failed
          - pexpect_dependency_absence.msg is contains 'pexpect'
```

LMK sometime if you want more info on when/why things *do* need to be quoted in YAML.

nitzmahone commented on test/integration/targets/expect/tasks/main.yml:
```diff
@@ -241,3 +241,106 @@
+    - name: assert failure with out of date pexpect
+      assert:
+        that:
+        - "pexpect_version_out_of_date.failed"
```
same quotes thing as above

nitzmahone commented on test/integration/targets/expect/tasks/main.yml:
```diff
@@ -241,3 +241,106 @@
+    - name: assert pexpect dependency absence
+      assert:
+        that:
+          - "pexpect_dependency_absence.failed"
```
This works-ish, but would potentially fail in an unexpected way or mask the real issue if, e.g. the `failed` key was not present. Sending the result to the `failed` Jinja test would be more robust e.g.: `pexpect_dependency_absence is failed`.

nitzmahone commented on test/integration/targets/expect/tasks/main.yml:
```diff
@@ -241,3 +241,106 @@
+
+    - name: Install pexpect3.3
+      pip:
+        break_system_packages: true
```
Curious if this was added to fix something? `break_system_packages` shouldn't be necessary here- even if the original venv was created to share system site packages (which it shouldn't be in this case anyway), a new installation would just mask it in the venv, not replace it.

nitzmahone commented on test/integration/targets/expect/tasks/main.yml:
```diff
@@ -241,3 +241,106 @@
     - "non_zero_result.msg == 'non-zero return code'"
+
+- name: run tests in venv
+  vars:
```
Not necessary, but if you want to get fancy, you could actually add `module_defaults` to this block to avoid passing `virtualenv` to all the `pip` module invocations...

nitzmahone commented on test/integration/targets/expect/tasks/main.yml:
```diff
@@ -241,3 +241,106 @@
+
+    - name: uninstall pexpect3.3
+      pip:
+        break_system_packages: true
```
IIRC the default pip module behavior is idempotent exact version match, so the explicit `absent` tasks between version changes should be unnecessary- it should just "make it so" for whatever version you specified.

nitzmahone reviewed (changes requested):
A few minor notes for the tests, but looks good- thanks!

pkingstonxyz committed:
Update tests according to corrections

pkingstonxyz committed:
Remove break_system_packages when installing pexpect 3.3

webknjaz commented on lib/ansible/modules/expect.py:
```diff
@@ -249,7 +249,7 @@ def main():
     delta = end_date - start_date
 
-    if b_out is None:
+    if not b_out:
```
Are there any other cases than `None` that we'd want to normalize? If it's expected to be `bytes | None`, then it's already either `b''` or `None`. And we don't really need to rewrite `b''` into `b''`.

webknjaz replied:
Another thing to consider: `pexpect.run()` is supposed to return `str`, not `bytes`. It must've been `bytes` in Python 2.

It's probably a good idea to see what's actually happening in runtime and see if correct string-ish types are being used in relation to this piece of data.

https://pexpect.readthedocs.io/en/stable/api/pexpect.html#run-function

pkingstonxyz replied:
I believe this whole condition could actually be removed because the value of b_out from [pexpect.run()](https://github.com/pexpect/pexpect/blob/master/pexpect/run.py) shouldn't ever return None. Changing the condition to be truthy means that the test coverage shows that branch of the if statement being covered while leaving the (probably unnecessary) guardrail in place.

pkingstonxyz committed:
Remove unnecessary if statement

pexpect.run(...) shouldn't ever return None, so it's safe
to remove the check entirely. Maintains 100% coverage
